src/drivers/imu/mpu9250/mag.cpp

```diff
@@ -209,14 +209,14 @@ MPU9250_mag::_measure(struct ak8963_regs data)
 
 	/*
 	 * Align axes - note the accel & gryo are also re-aligned so this
-	 *              doesn't look obvious with the datasheet
+	 *              doesn't look obvious without the datasheet
 	 */
-	mrb.x_raw =  data.x;
-	mrb.y_raw = -data.y;
+	mrb.x_raw =  data.y;
+	mrb.y_raw =  data.x;
 	mrb.z_raw = -data.z;
 
-	float xraw_f =  data.x;
-	float yraw_f = -data.y;
+	float xraw_f =  data.y;
```

Comment:
Any idea what was going on here? I wasn't sure if it was an attempt to conform with other external mags like 3dr unit.
@mhkabir @dakejahl

Reply:
The idea behind this is to bring the mag into the same reference frame as the accel and gyro. 

![image](https://user-images.githubusercontent.com/2019539/40027990-c857e9ca-5799-11e8-9a0c-263d3891e51b.png)

Reply:
I meant the existing mapping that clearly doesn't match the datasheet. This is the common orientation for external mags. https://github.com/PX4/Firmware/blob/master/src/drivers/magnetometer/hmc5883/hmc5883.cpp#L980-L983

Reply:
@dagar It was designed to match the orientation on Pixhawk. Not necessarily the best approach in terms of maintainability.

Reply:
To fix the driver for the generic case it's going to force each user to reset CAL_MAG0_ROT and recalibrate.  It seems important that this driver should be corrected to match the datasheet, both for the temperature and for the orientation.  Is it possible to slate this for the 2.0.0 release and (unfortunately) force users who upgrade to accept the backwards breakage for the mpu9250 along the lines of what I have suggested to #9054?

Reply:
Something still doesn't make sense here. How is this functional at all for existing pixracer or pixhawk 2.1 users? In both cases they have an internal mpu9250 no additional mag rotation applied.

Reply:
@dagar It looks like the calibration routine is applying negative scalars to the values and making it work.

Reply:
It works when the calibration applies negative scalars.

Reply:
Our mpu9250 sensor board orientation matches pixracer precisely.

Reply:
I don't see how negative scale factors would account for swapping x and y without an additional rotation error, but then why aren't accel and gyro a problem? Something else would have to be wrong for this to make sense.


Here's a stock pixracer with default rotation.

`sensors status` level facing south
![screenshot from 2018-05-21 20-32-33](https://user-images.githubusercontent.com/84712/40336278-bf87b080-5d36-11e8-9e1d-976c378d9475.png)

`sensors status` nose down
![screenshot from 2018-05-21 20-33-21](https://user-images.githubusercontent.com/84712/40336290-d2c17e4c-5d36-11e8-97f8-db096b274b3c.png)

mag calibration (no negative scale factors)
![screenshot from 2018-05-21 20-34-04](https://user-images.githubusercontent.com/84712/40336304-dfb07fcc-5d36-11e8-8948-71a871235487.png)

No issues with gyro or accel inconsistencies either.

Reply:
@dagar, see my edited comment above. You are correct in that it indeed requires a change to the driver startup rotation to correct the system in its entirety.